go-runtime/ftl/ftltest/fake.go

```diff
@@ -121,3 +181,236 @@ func actuallyCallMap(ctx context.Context, impl mapImpl) any {
 	}
 	return out
 }
+
+func (f *fakeFTL) PublishEvent(ctx context.Context, topic *schema.Ref, event any) error {
+	f.publishWaitGroup.Add(1)
+	f.globalTopic.PublishSync(publishEvent{topic: topic, content: event})
+	return nil
+}
+
+// addSubscriber adds a subscriber to the fake FTL instance. Each subscriber included in the test must be manually added
+func addSubscriber[E any](f *fakeFTL, sub ftl.SubscriptionHandle[E], sink ftl.Sink[E]) {
+	f.pubSubLock.Lock()
+	defer f.pubSubLock.Unlock()
+
+	if _, ok := f.subscriptions[sub.Name]; !ok {
+		f.subscriptions[sub.Name] = &subscription{
+			name:   sub.Name,
+			topic:  sub.Topic,
+			errors: map[int]error{},
+		}
+	}
+
+	subscribers, ok := f.subscribers[sub.Name]
+	if !ok {
+		subscribers = []subscriber{}
+	}
+	f.subscribers[sub.Name] = append(subscribers, func(ctx context.Context, event any) error {
+		if event, ok := event.(E); ok {
+			return sink(ctx, event)
+		}
+		return fmt.Errorf("unexpected event type %T for subscription %s", event, sub.Name)
+	})
+}
+
+// eventsForTopic returns all events published to a topic
+func eventsForTopic[E any](ctx context.Context, f *fakeFTL, topic ftl.TopicHandle[E]) []E {
+	// Make sure all published events make it into our pubsub state
+	f.publishWaitGroup.Wait()
+
+	f.pubSubLock.Lock()
+	defer f.pubSubLock.Unlock()
+
+	logger := log.FromContext(ctx).Scope("pubsub")
+	var events = []E{}
+	raw, ok := f.topics[topic.Ref.ToRefKey()]
+	if !ok {
+		return events
+	}
+	for _, e := range raw {
+		if event, ok := e.(E); ok {
+			events = append(events, event)
+		} else {
+			logger.Warnf("unexpected event type %T for topic %s", e, topic.Ref)
+		}
+	}
+	return events
+}
+
+// resultsForSubscription returns all consumed events and whether an error was returned
+func resultsForSubscription[E any](ctx context.Context, f *fakeFTL, handle ftl.SubscriptionHandle[E]) []SubscriptionResult[E] {
+	f.pubSubLock.Lock()
+	defer f.pubSubLock.Unlock()
+
+	logger := log.FromContext(ctx).Scope("pubsub")
+	results := []SubscriptionResult[E]{}
```

Comment:
```suggestion
	var results []SubscriptionResult[E]
```

tip: avoids an allocation